libs/wingc/src/struct_schema.rs

```diff
@@ -0,0 +1,185 @@
+
+				code.close("}");
+				code.to_string().strip_suffix("\n").unwrap().to_string()
+			}
+			Type::Map(ref t) => {
+				let mut code = CodeMaker::default();
+				code.open("{");
+
+				code.line("type: \"object\",");
+				code.line(format!(
+					"patternProperties: {{ \".*\": {} }}",
+					self.jsify_struct_schema_field(t, env)
+				));
+
+				code.close("}");
+				code.to_string().strip_suffix("\n").unwrap().to_string()
+			}
+			Type::Optional(t) => self.jsify_struct_schema_field(&t, env),
+			Type::Json(_) => "{ type: \"object\" }".to_string(),
+			_ => "{ type: \"null\" }".to_string(),
+		}
+	}
+
+	fn jsify_struct_env_properties(&self, env: &SymbolEnv) -> CodeMaker {
+		let mut code = CodeMaker::default();
+		for (field_name, (.., kind)) in env.symbol_map.iter() {
+			code.line(format!(
+				"{}: {},",
+				field_name,
+				self.jsify_struct_schema_field(&kind.as_variable().unwrap().type_, env)
+			));
+		}
+		code
+	}
+
+	fn jsify_struct_schema(&self, struct_: &Struct, _env: &SymbolEnv) -> CodeMaker {
+		let mut code = CodeMaker::default();
+
+		code.open("module.exports = function(stdStruct) {".to_string());
+		code.open(format!("class {} {{", struct_.name));
+
+		code.open("static jsonSchema() {".to_string());
+		code.open("return {");
+
+		code.line(format!("id: \"/{}\",", struct_.name));
+		code.line("type: \"object\",".to_string());
+
+		code.open("properties: {");
+
+		code.add_code(self.jsify_struct_env_properties(&struct_.env));
+
+		//close properties
+		code.close("},");
+
+		code.add_code(self.jsify_struct_schema_required_fields(&struct_.env));
+
+		// close return
+		code.close("}");
+
+		// close schema
+		code.close("}");
+
+		// create _validate() function
+		code.open("static fromJson(obj) {");
+		code.line("return stdStruct._validate(obj, this.jsonSchema())");
+		code.close("}");
+
+		// create _toInflightType function that just requires the generated struct file
+		code.open("static _toInflightType(context) {".to_string());
+		code.line(
+			"return `require(\"./${require('path').basename(__filename)}\")(${ context._lift(stdStruct) })`;".to_string(),
+		);
+		code.close("}");
+
+		// close class
+		code.close("}");
+		// close module.exports
+
+		code.line(format!("return {};", struct_.name.name));
+		code.close("};");
+
+		code
+	}
+}
+
+// Looks for any references to a struct type, and emits a struct schema file for it.
+impl<'a> Visit<'a> for StructSchemaVisitor<'a> {
+	fn visit_reference(&mut self, node: &'a crate::ast::Reference) {
+		match node {
+			Reference::InstanceMember { .. } => {
+				visit::visit_reference(self, node);
+			}
+			Reference::TypeMember { type_name, .. } => {
+				if is_udt_struct_type(type_name, self.ctx.current_env().unwrap()) {
+					let type_ = resolve_user_defined_type(type_name, self.ctx.current_env().unwrap(), 0);
+					let struct_code = self.jsify_struct(type_.unwrap().as_struct().unwrap());
+					self.jsify.emit_struct_file(
```

Comment:
Is `emit_struct_file` only called from here? If so maybe move it out of `jsify` and drop the jsifier dependency for this visitor. (Only have a `&types` dependency?)